### PR TITLE
Add specialized checker for firefox + some small improvements

### PR DESCRIPTION
**Pull request by andrunko**
https://phabricator.endlessm.com/T21208

### CODE REVIEW

**Comment by andrunko** [comment]
Force pushed to add `--rm` to `podman run` in the `run-in-container.sh` script.

File: src/flatpak-external-data-checker [hunk 1]
```diff
@@ -183,6 +183,10 @@ def main():
                         help="Update manifest(s) to refer to new versions of "
                              "external data",
                         action="store_true")
+    parser.add_argument("--commit-only",
+                        help="Do not open PRs for updates, only commit changes "
+                             "to external data (requires --update)",
```
Comment (wjt) [inline src/flatpak-external-data-checker]: I think it should imply `--update`, rather than documenting that it does nothing unless you pass another flag.

File: src/lib/externaldata.py [hunk 2]
```diff
@@ -74,9 +74,9 @@ def __str__(self):
                '  Checker: {checker_data}'.format(state=self.state.name,
                                                   filename=self.filename,
                                                   type=self.type.name,
-                                                  url=self.url,
-                                                  checksum=self.checksum,
-                                                  size=self.size,
+                                                  url=self.new_version.url if self.new_version else self.current_version.url,
```
Comment (wjt) [inline src/lib/externaldata.py]: Personally I would prefer (eg):

```python3
version = self.new_version or self.current_version
...
    url=version.url,
    checksum=version.checksum,
...
```

but, not a big deal.

File: src/lib/externaldata.py [hunk 3]
```diff
@@ -137,36 +137,41 @@ def update(self):
...
-            for i, arg in enumerate(finish_args)
+            cls(source_path, arg, finish_args)
+            for arg in finish_args
             if arg.startswith(cls.PREFIX)
         ]
 
     def update(self):
         if self.new_version is not None:
+            if not self.source in self.source_parent:
+                log.warning("Unable to find index for %s", self.source)
```
Comment (wjt) [inline src/lib/externaldata.py]: This should only happen on programmer error – I'd just leave this check out and allow `index()` to throw an exception.

File: tests/test_checker.py [hunk 4]
```diff
@@ -69,7 +69,9 @@ def test_check_filtered(self):
 
         ext_data = dummy_checker.check()
         ext_data_from_getter = dummy_checker.get_external_data()
-        self.assertEqual(ext_data_from_getter, ext_data)
+        self.assertEqual(len(ext_data), len(ext_data_from_getter))
+        for data in ext_data:
+            self.assertIn(data, ext_data_from_getter)
```
Comment (wjt) [inline tests/test_checker.py]: Why is this change needed – is the order no longer preserved?

File: src/checkers/firefoxchecker.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        for line in sha256_table.splitlines():
+            sha256, path = line.split(None, 1)
+            path = path.strip()
+
+            if path.startswith(self.FIREFOX_PLATFORM):
+                assert path.count('/') == 2, path
+                _, dirname, filename = path.split('/')
+
+                source_filename = None
+                if re.match(checker_data['browser']['filename-match'], filename) and \
```
Comment (wjt) [inline src/checkers/firefoxchecker.py]: Why do we need to specify regular expressions to match in the manifest? This checker is Firefox-specific. We can just hardcode them.

File: src/checkers/firefoxchecker.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+                }
+                # add new data to the same manifest as the main browser tarball
+                data = ExternalDataSource.from_source(browser_data.source_path, source, browser_data.source_parent)
+                data.state = ExternalData.State.ADDED
+                added.append(data)
+
+        return added
+
+    def _get_latest_available_version(self):
+        if self._latest_version:
```
Comment (wjt) [inline src/checkers/firefoxchecker.py]: I think you can drop all the manual caching and decorate this function with [functools.lru_cache()](https://docs.python.org/3/library/functools.html#functools.lru_cache):

```python3
    @functools.lru_cache()
    def _get_latest_available_version(self):
```

File: src/checkers/firefoxchecker.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+        latest_version_data = {}
+        request_url = self.FIREFOX_RELEASE_INFO_URL
+        with urllib.request.urlopen(request_url) as resp:
+            latest_version_data = json.load(resp)
+
+        if isinstance(latest_version_data, dict) and 'LATEST_FIREFOX_VERSION' in latest_version_data:
+            self._latest_version = latest_version_data['LATEST_FIREFOX_VERSION']
+        else:
+            self._latest_version = None
```
Comment (wjt) [inline src/checkers/firefoxchecker.py]: We should fail in this case.

**Review by wjt — CHANGES_REQUESTED** [review]

File: tests/test_checker.py (hunk 4 above)
Comment (andrunko) [reply tests/test_checker.py]: Yes, the order is not preserved on the return value of `ManifestChecker.check()` anymore. That is because we run the modules check first and add all modules external data to the check list and then we check individual sources that were not processed yet and add them as they get processed (opposed to only checking sources in order as it was before).

As far as I could see nothing (apart from this test) was using the return value of `ManifestChecker.check()`, so I think it should be safe but please let me know.

File: src/checkers/firefoxchecker.py (hunk 7 above)
Comment (andrunko) [reply src/checkers/firefoxchecker.py]: This method is invoked inside a `try/catch/finally` and the `self._latest_version` is checked on `finally` with a warning if set to `None`.

Do you think we should raise an exception (i.e. break the checker) in this case? Or how else 
 would you fail here?

File: src/checkers/firefoxchecker.py (hunk 5 above)
Comment (andrunko) [reply src/checkers/firefoxchecker.py]: Ok, tbh I prefer to have it configurable such that we don't have to force sources/translations filename to match (e.g. in a future official firefox flatpak). 

Maybe leave the checker options but use default values if they are not set?

File: src/lib/externaldata.py (hunk 3 above)
Comment (andrunko) [reply src/lib/externaldata.py]: Ack.

File: src/flatpak-external-data-checker (hunk 1 above)
Comment (andrunko) [reply src/flatpak-external-data-checker]: Ack.

File: src/checkers/firefoxchecker.py (hunk 6 above)
Comment (andrunko) [reply src/checkers/firefoxchecker.py]: Ok, but isn't this more expensive than the current cache? Not that we care much about resources here but still...

File: src/checkers/firefoxchecker.py (hunk 7 above)
Comment (wjt) [reply src/checkers/firefoxchecker.py]: > Do you think we should raise an exception (i.e. break the checker) in this case?

Probably! If this web service turns out to be unreliable we can do something different.

File: src/checkers/firefoxchecker.py (hunk 6 above)
Comment (wjt) [reply src/checkers/firefoxchecker.py]: It'll be a dictionary whose key is `self` and whose value is the return value – a dozen bytes or so. I think this is worth the code size reduction.

File: src/checkers/firefoxchecker.py (hunk 5 above)
Comment (wjt) [reply src/checkers/firefoxchecker.py]: A future official Firefox Flatpak will be generated as part of the Mozilla release process so will not need this script.

Comment (wjt) [reply src/checkers/firefoxchecker.py]: > Maybe leave the checker options but use default values if they are not set?

OK! I predict that we'll never need to set them but I am prepared to be wrong :-)

File: tests/test_checker.py (hunk 4 above)
Comment (wjt) [reply tests/test_checker.py]: It's probably fine but perhaps you could express this as:

```
         self.assertEqual(len(ext_data), len(ext_data_from_getter))
         self.assertEqual(set(ext_data), set(ext_data_from_getter))
```

?

File: src/flatpak-external-data-checker (hunk 1 above)
Comment (andrunko) [reply src/flatpak-external-data-checker]: Done.

File: src/lib/externaldata.py (hunk 2 above)
Comment (andrunko) [reply src/lib/externaldata.py]: Done.

File: src/lib/externaldata.py (hunk 3 above)
Comment (andrunko) [reply src/lib/externaldata.py]: Done.

File: tests/test_checker.py (hunk 4 above)
Comment (andrunko) [reply tests/test_checker.py]: Done.

File: src/checkers/firefoxchecker.py (hunk 5 above)
Comment (andrunko) [reply src/checkers/firefoxchecker.py]: Done, ended up hardcoding the expressions in the checker - if in the future we find the need we can make it configurable again.

File: src/checkers/firefoxchecker.py (hunk 6 above)
Comment (andrunko) [reply src/checkers/firefoxchecker.py]: Done.

File: src/checkers/firefoxchecker.py (hunk 7 above)
Comment (andrunko) [reply src/checkers/firefoxchecker.py]: Done.

**Comment by andrunko** [comment]
Everything should be updated, please check again.

**Comment by andrunko** [comment]
@wjt Squashed the fixup for the script permissions, thanks!

File: src/checkers/firefoxchecker.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+        log.debug('Retrieving latest firefox version info')
+        latest_info = None
+        try:
+            latest_info = self._get_latest_version_info(latest_version)
+        except Exception as e:
+            log.warning('Error retrieving latest firefox version info: %s', e)
+        finally:
+            if not latest_info:
+                log.warning('Could not retrieve latest firefox version information')
```
Comment (wjt) [inline src/checkers/firefoxchecker.py]: Again I would prefer to skip the exception handling and just let it bubble up in this case. We won't notice the warnings, but we will notice the job failing.

File: src/checker.py [hunk 9]
```diff
@@ -189,15 +227,24 @@ def get_outdated_external_data(self):
...
         for data in datas:
-            if data.new_version is None:
+            if data.new_version is None and \
+               data.state != ExternalData.State.ADDED and \
+               data.state != ExternalData.State.REMOVED:
                 continue
 
             data.update()
+            if data.state == ExternalData.State.ADDED:
+                message = "Added {}".format(data.filename)
```
Comment (wjt) [inline src/checker.py:245]: I guess we'll find out whether, in practice, new translations are added/removed within the lifetime of a given Firefox release. I would expect not.

It would be nice if the commit message could read “Update Firefox to X.Y.Z” rather than “Update 94 modules”; and this would also make #17 easier because we would have already boiled the update down to the salient fact. But, as Voltaire probably didn't say, RESOLVED LATER is the friend of good!

Each time Firefox is updated, the commit message will be 180 lines long because every file will be both added (new version) and removed (old version) whereas the logical change is "Update Firefox to ____". I feel like it would be nice to be able to do

**Review by wjt — APPROVED** [review]
I had a few other comments but I just fixed them directly.

File: src/checker.py (hunk 9 above)
Comment (andrunko) [reply src/checker.py:245]: Makes sense, I will have a think about it - we could probably use the module the updated data comes from to build the commit msg.
